// @dendronhq/dendron-cli/lib/src/commands/__tests__/build-site-v2.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_server_1 = require("@dendronhq/common-server");
const common_test_utils_1 = require("@dendronhq/common-test-utils");
const engine_server_1 = require("@dendronhq/engine-server");
const engine_test_utils_1 = require("@dendronhq/engine-test-utils");
const build_site_v2_1 = require("../build-site-v2");
// TODO
describe("basic", () => {
    let siteRootDir;
    beforeEach(async () => {
        siteRootDir = common_server_1.tmpDir().name;
    });
    test.skip("basic", async () => {
        await common_test_utils_1.runEngineTestV4(async ({ wsRoot }) => {
            const cmd = new build_site_v2_1.BuildSiteV2CLICommand();
            const resp = await cmd.eval({ wsRoot, stage: "dev", serve: false });
            return resp;
        }, {
            createEngine: engine_server_1.createEngine,
            expect,
            preSetupHook: async (opts) => {
                await common_test_utils_1.ENGINE_HOOKS.setupBasic(opts);
                await engine_test_utils_1.ConfigUtils.writeConfig({
                    config: {
                        version: 0,
                        vaults: opts.vaults,
                        site: { siteHierarchies: ["foo", "bar"], siteRootDir },
                    },
                    wsRoot: opts.wsRoot,
                });
            },
        });
    });
});
