// @dendronhq/dendron-cli/lib/src/commands/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLICommand = exports.BaseCommand = void 0;
const common_server_1 = require("@dendronhq/common-server");
const lodash_1 = __importDefault(require("lodash"));
class BaseCommand {
    constructor(name, opts) {
        opts = lodash_1.default.defaults(opts, { log2Stdout: false });
        this.L = common_server_1.createLogger(name || "Command");
    }
}
exports.BaseCommand = BaseCommand;
class CLICommand extends BaseCommand {
    constructor(opts) {
        super(opts.name, opts);
        this.eval = async (args) => {
            this.L.info({ args });
            const opts = await this.enrichArgs(args);
            return this.execute(opts);
        };
        this.name = opts.name;
        this.desc = opts.desc;
    }
    buildArgs(args) {
        args.option("wsRoot", {
            describe: "location of workspace",
            demandOption: true,
        });
    }
    buildCmd(yargs) {
        return yargs.command(this.name, this.desc, this.buildArgs, this.eval);
    }
}
exports.CLICommand = CLICommand;
