// @dendronhq/engine-server/lib/markdown/remark/__tests__/dendronPub.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_test_utils_1 = require("@dendronhq/common-test-utils");
const lodash_1 = __importDefault(require("lodash"));
const enginev2_1 = require("../../../enginev2");
const types_1 = require("../../types");
const utils_1 = require("../../utils");
function proc(engine, dendron, opts) {
    return utils_1.MDUtilsV4.procFull({ engine, ...dendron, publishOpts: opts });
}
describe("basics", () => {
    let engine;
    let dendronData = {
        dest: types_1.DendronASTDest.HTML,
    };
    test("imagePrefix", () => {
        const out = proc(engine, dendronData, {
            assetsPrefix: "bond/",
        }).processSync(`![alt-text](image-url.jpg)`);
        expect(lodash_1.default.trim(out.toString())).toEqual("![alt-text](/bond/image-url.jpg)");
    });
    test("imagePrefix2", () => {
        const out = proc(engine, dendronData, {
            assetsPrefix: "/bond/",
        }).processSync(`![alt-text](/image-url.jpg)`);
        expect(lodash_1.default.trim(out.toString())).toEqual("![alt-text](/bond/image-url.jpg)");
    });
    test("can't publish", async () => {
        await common_test_utils_1.runEngineTestV4(async ({ engine, vaults }) => {
            const vault = vaults[0];
            const config = {
                siteHierarchies: ["foo"],
                siteRootDir: "foo",
            };
            const resp = await utils_1.MDUtilsV4.procRehype({
                proc: proc(engine, { ...dendronData, config, vault }, {
                    wikiLinkOpts: { useId: true },
                    transformNoPublish: true,
                }),
            }).process(`[[an alias|bar]]`);
            expect(resp).toMatchSnapshot();
            expect(resp.contents).toEqual(`<p><a href="403.html">an alias</a></p>`);
        }, { expect, createEngine: enginev2_1.createEngine, preSetupHook: common_test_utils_1.ENGINE_HOOKS.setupBasic });
    });
});
